README.md
home-website
============
Copyright SJ Hindmarch 2013, 2014
Distributed under the GNU General Public License
============

Stuff I am doing at home learning about HTML5, CSS and JavaScript




ChangeLog
2015-01-20  Steve  <sjh@kanga.hindmarch.home>

	* navbar.html: 
	- adding dropdowns to the nav bar

2014-11-30  Steve  <sjh@kanga.hindmarch.home>

	* index.html
	* index.css
	* navbar.html
	* bootstrap/index.html
	- added an introduction to bootstrap

2014-01-09  Steve Hindmarch  <stephen.hindmarch@bt.com>

	* d3_iwd_v3.html:
	* d3_iwd_v1.js:
	* d3_iwd_v3.js:
	- working on a delete function 

2014-01-05  Steve  <sjh@kanga.hindmarch.home>

	* d3_iwd_v3.html:
	* d3_iwd_v3.js:
	* index.css:
	* iwd.css:
	- finished the profiles
	- added the rest of the priorities
	- added sections and footers with styling

	* ChangeLog
	- project change log created - better late than never eh?



d3_bar.js
/* 
 * D3 Bar Chart Example
 * Copyright SJ Hindmarch 2013, 2014
 * Distributed under the GNU General Public License
 */

function init(){
    var dataset=[20,24,32,45,32,21,65,78,101,102,54,156,44,12,18,99];
    var svg=d3.select('#workarea').append('svg');
    process(svg,dataset);
}

function new_chart(){
    var s_data=d3.select('#indata').property('value').split(" ");
    var i_data=s_data.map(function(x){return parseInt(x,10)});
    
    d3.select('#workarea').select('svg').remove();
    var svg=d3.select('#workarea').append('svg').attr('class','chart');
    process(svg,i_data);
}

function process(svg,dataset){
    var h=180;
    var w=360;
    var bar_padding=1;

    var bar_space=w/dataset.length;
    var bar_width=bar_space-bar_padding;
    var d_max=dataset.slice().sort(function(a,b){return b-a})[0];
    var y_scale=d3.scale.linear().domain([0,d_max]).range([h,0])

    svg.selectAll('rect').data(dataset).enter().append('rect')
    .attr('x',function(d,i){return i*bar_space})
    .attr('y',function(d,i){return y_scale(d)})
    .attr('width',bar_width)
    .attr('height',h)
    .attr('fill',function(d,i){return 'rgba(0,0,'+(255-d)+',1)'});

    svg.selectAll('text').data(dataset).enter().append('text')
    .text(function(d,i){return d})
    .attr('x',function(d,i){return i*bar_space+bar_width/2})
    .attr('y',function(d,i){return y_scale(d)+10})
    .attr('font-family','sans-serif')
    .attr('font-size', '11px')
    .attr('fill', 'white')
    .attr("text-anchor", "middle");
}


d3_joins.js
/* 
 * D3 Data Joins Example
 * Copyright SJ Hindmarch 2013, 2014
 * Distributed under the GNU General Public License
 */

var scale_x=d3.scale.linear().domain([0,100]).range([0,360]);
var scale_r=d3.scale.linear().domain([0,10]).range([0,90]);

function init(){
    d3.select('#workarea').append('p').text('Enter data and press return');
};

function process(){
    // capture data as strings and convert to integers
    var s_data=d3.select('#indata').property('value').split(" ");
    var i_data=s_data.map(function(x){return parseInt(x,10)});

    // Organise display
    var workarea=d3.select('#workarea');
    var bararea=workarea.selectAll('[name="bararea"]');
    if(bararea.empty()){
	bararea=workarea.append('div').attr('name','bararea');
	workarea.append('br');
    }
    var boxarea=workarea.selectAll('[name="boxarea"]');
    if(boxarea.empty()){
	boxarea=workarea.append('div').attr('name','boxarea');
    }

    // Show data in text area
    workarea.select('p').text(s_data.toString());

    // Barchart can be drawn direct to the page
    var bar=bararea.selectAll('.bar').data(i_data);
    bar.enter().append('div').attr('class','bar');
    bar.exit().remove();
    bar.style('height',function(d){return d+'px'});

    // Bar labels shown underneath
    var box=boxarea.selectAll('.box').data(i_data);
    box.enter().append('div').attr('class','box');
    box.exit().remove();
    box.text(function(d){return d});

    // Circle chart requires an svg
    var svg=workarea.selectAll('svg');
    if(svg.empty()){
	svg=workarea.append('svg');
    }
    var c=svg.selectAll('circle').data(i_data);
    c.enter().append("circle");
    c.attr('cy',90)
    .attr('cx',function(d){return scale_x(d)})
    .attr('r',function(d){return scale_r(Math.sqrt(d))});
    c.exit().remove();
}


d3_iwd_v1.js
/* 
 * iWD Priority Visualisation With D3
 * Copyright SJ Hindmarch 2013, 2014
 * Distributed under the GNU General Public License
 *
 * Basic chart drawing and priority flow modelling
 *
 */

/* Create a priority class that stores priority data */
function Priority(name,color,initialPriority,transitions){
    this.name=name;
    this.color=color;
    this.initP=initialPriority;
    this.transitions=transitions;
};

/* Function to turn the priority values into a line array */
Priority.prototype.getPriorityLine=function(startTime,endTime){
    var line=[];

    /* If transitions not set return flat line */
    if(this.transitions==null || this.transitions.length==0){
	line.push([startTime,this.initP]);
	line.push([endTime,this.initP]);
	return line;
    }

    /* Find start of line */
    var i=0;
    var currP=this.initP;
    while(line.length==0 && i<this.transitions.length){
	var p=this.transitions[i++];
	if(p[0]>startTime){
	    line.push([startTime,currP]);
	    currP=p[1];
	}
    }
    for(var j=i-1;j<this.transitions.length;j++){
	var p=this.transitions[j];
	if(p[0]<endTime){
	    line.push(p);
	    currP=p[1];
	}
    }
    line.push([endTime,currP]);
    
    return line;
};

/* Function to get a transition at a specific time */
Priority.prototype.getTransition=function(time){
    if(this.transitions==null || this.transitions.length==0){
	return null;
    }

    for(var i=0;i<this.transitions.length;i++){
	var transition=this.transitions[i]
	if(transition[0]==time){
	    return transition[1];
	}
    }
    return null;
};

/* Create a chart class that can draw a basic chart */
function PriorityChart(chartName,tableName){
    /* Store the priority objects */
    this.priorities={};

    /* Set dimensions of area we will be drawing in */
    this.h=480;
    this.w=800;
    this.padding=10;
    this.x_padding=40;

    /* Select graphical area  */
    this.g=d3.select(chartName);
    this.table=d3.select(tableName);

    /* Create drawing scales */
    this.dx_min=-300;
    this.dx_max=150;
    this.dy_min=0;
    this.dy_max=4200;

    this.x_scale=d3.scale.linear().domain([this.dx_min,this.dx_max])
    .range([this.padding,this.w-this.padding])
    this.y_scale=d3.scale.linear().domain([this.dy_min,this.dy_max])
    .range([this.h-this.x_padding,this.padding])

    /* Create axes */
    this.x_axis=d3.svg.axis().scale(this.x_scale);
    this.y_axis=d3.svg.axis().scale(this.y_scale).orient('left');

    /* Create line drawing generator */
    this.lineG=d3.svg.line()
    .x(function(d){return this.x_scale(d[0])})
    .y(function(d){return this.y_scale(d[1])})
    .interpolate("step-after");

    this.drawChart();
};

/* Add a method to draw the empty chart */
PriorityChart.prototype.drawChart=function(){
    /* Expand drawing area to required size */
    this.g.style('width',this.w).style('height',this.h);
    this.drawGrid();
    this.drawAxes();
};

/* Add a method to draw the grid */
PriorityChart.prototype.drawGrid=function(){
    /* x grid */
    this.g.append('g').attr('class','axis').attr('id','x_axis')
    .attr('transform','translate(0,'+(this.h - this.x_padding)+')')
    .call(this.x_axis);

    /* y grid */
    this.g.append('g').attr('class','grid')
    .attr('transform', 'translate('+this.padding+',0)')
    .call(
	this.y_axis.tickSize((2*this.padding)-this.w,0,0).tickFormat('')
	);
}

/* Add a method to draw the axes */
PriorityChart.prototype.drawAxes=function(){

    /* x axis */
    this.g.append('g').attr('class','grid')
    .attr('transform', 'translate(0,'+this.y_scale(this.dy_max)+')')
    .call(
	this.x_axis.tickSize(
	    this.y_scale(this.dy_min)-this.y_scale(this.dy_max),0,0).
	tickFormat('')
	);

    /* y axis */
    this.g.append('g').attr('class','axis').attr('id','y_axis')
    .attr('transform','translate('+this.x_scale(0)+',0)')
    .call(this.y_axis.tickSize(5,0,0).tickFormat(d3.format('>4d')));

};

/* Draw the label for an individual priority */
PriorityChart.prototype.drawLabel=function(item){
    this.g.append("text").text(item.name).attr('label',item.name)
    .attr("x",this.x_scale(this.dx_min))
    .attr("y",this.y_scale(item.initP))
    .attr("text-anchor","start").attr("dy","-0.5em")
    .attr("font-size","75%");
};

/* Draw the graphics for an individual priority */
PriorityChart.prototype.drawLine=function(item){
    /* Get the priority line from the priority item */
    var lineD=item.getPriorityLine(this.dx_min,this.dx_max);

    /* Draw the line */
    this.g.append("path").attr('label',item.name)
    .attr('stroke-width',2).attr('fill','none')
    .attr('stroke',item.color)
    .attr('d',this.lineG(lineD));

};

/* Add a priority to the chart and draw it */
PriorityChart.prototype.addPriority=function(){
    var chart=this;
    for(var i=0;i<arguments.length;i++){
	var item=arguments[i];
	chart.priorities[item.name]=item;
	chart.drawLabel(item);
	chart.drawLine(item);
    }
};

PriorityChart.prototype.updatePriority=function(){
    var chart=this;
    for(var i=0;i<arguments.length;i++){
	var item=arguments[i];
	//console.log('PriorityChart::updatePriority:'+item.name);
	this.g.selectAll("[label='"+item.name+"']").remove();
	chart.priorities[item.name]=item;
	chart.drawLabel(item);
	chart.drawLine(item);
    }
};
    
PriorityChart.prototype.outputData=function(){
    var header=this.table.insert('tr');
    header.append('th').text('Name');
    header.append('th').text('Colour');
    header.append('th').text('Initial Priority');
    header.append('th').text('Transitions');
    for(var key in this.priorities){
	var item=this.priorities[key];
	var row=this.table.insert('tr');
	row.insert('td').text(item.name).attr('class','name');
	row.insert('td').text(item.color).attr('class','color');
	row.insert('td').text(item.initP).attr('class','initp');
	var tRow=row.insert('td').attr('class','priorities');
	if(item.transitions){
	    tText="";
	    item.transitions.forEach(function(x){
		    tText+=" "+x[1]+"@"+x[0]+";";
		});
	    tRow.text(tText);
	}else{
	    tRow.text("-").style('text-align','center');
	}
	    
    }
}


d3_iwd_v2.js
/* 
 * iWD Priority Visualisation With D3
 * Copyright SJ Hindmarch 2013, 2014
 * Distributed under the GNU General Public License
 *
 * Create priority objects by reading from a table
 *
 */

/* Create a child priority class that gets its values from a table row */
function SimpleRowPriority(row){
    this.row=row;
    var name=row.select('.name').text();
    var color=row.select('.color').text();
    var initP=row.select('.initp').node().value;
    var transitions=this.scanTransitions(row);
    Priority.call(this,name,color,initP,transitions);
}

/* Set up inheritance */
SimpleRowPriority.prototype=new Priority();
SimpleRowPriority.constructor=SimpleRowPriority;

/* Add function to process the transitions */
SimpleRowPriority.prototype.scanTransitions=function(row){
    var transitions=[];
    row.selectAll('input.transition').each(function(d,i){
	var item=d3.select(this);
	var time=item.attr('time');
	var value=item.node().value;
	if(value!=""){
	    transitions.push([time,value]);
	}
    });
    return transitions;
}

/* A table scanning function. Pass in table name and 
 * type of priority to be scanned for.
 */
function scanPrioritiesTable(tableName,type){
    rows=d3.select(tableName).selectAll('tr.priority');
    priorities=[];
    rows.each(function(d,i){
	priorities.push(new type(d3.select(this)));
    });
    return priorities;
}

/* Create the table from initial data */
function createSimpleRowPrioritiesTable(tableId,priorities){
    var times=['-120','-60','-30','0','30','60','120'];

    var table=d3.select(tableId);
    priorities.forEach(function(priority){
	    var row=table.append('tr').attr('class','priority');
	    row.append('td').attr('class','name').text(priority.name);
	    row.append('td').attr('class','color').text(priority.color);
	    row.append('td').append('input').attr('type','text')
            .attr('class','initp').attr('maxlength','4').attr('size','8')
	    .attr('value',priority.initP);
	    times.forEach(function(d,i){
		    var value=priority.getTransition(d);
		    var cell=row.append('td').append('input')
		    .attr('type','text')
		    .attr('class','transition').attr('maxlength','4')
		    .attr('size','8').attr('time',d)
		    if(value){
			cell.attr('value',value);
		    }
		});
	});
    
}

/* Add a row to the table */
function addSimplePriorityRow(tableId,nameId,colorId,initpId){
    var times=['-120','-60','-30','0','30','60','120'];

    var name=d3.select(nameId).node().value;
    var color=d3.select(colorId).node().value;
    var initp=d3.select(initpId).node().value;

    var table=d3.select(tableId);
    var row=table.append('tr').attr('class','priority');
    row.append('td').attr('class','name').text(name);
    row.append('td').attr('class','color').text(color);
    row.append('td').append('input').attr('type','text').attr('class','initp')
    .attr('maxlength','4').attr('size','8').attr('value',initp);
    times.forEach(function(x){
	    row.append('td').append('input').attr('type','text')
	    .attr('class','transition').attr('maxlength','4').attr('size','8')
	    .attr('time',x);
	});
}

/* redraw the chart from the current table */
function redrawPriorityChart(chart,tableId,pFunction){
    var priorities=scanPrioritiesTable(tableId,pFunction);
    console.log('redraw: priorities.length='+priorities.length);
    priorities.forEach(function(x){
	    //console.log('redraw: priority: '+x.name+','+x.profile);
	    chart.updatePriority(x);
	});
}


d3_xyplot.js
/* 
 * D3 XY Plot Example
 * Copyright SJ Hindmarch 2013, 2014
 * Distributed under the GNU General Public License
 */

function init(){
    var dataset=[[0,0],[10,25],[60,42],[94,97],[99,27],[200,100]];

    var h=180;
    var w=360;
    var ax_padding=20;
    var ay_padding=30;
    var padding=10;

    var dx_max=dataset.slice().sort(function(a,b){return b[0]-a[0]})[0][0];
    var dy_max=dataset.slice().sort(function(a,b){return b[1]-a[1]})[0][1];
    var x_scale=d3.scale.linear().domain([0,dx_max])
    .range([ay_padding,w-padding])
    var y_scale=d3.scale.linear().domain([0,dy_max])
    .range([h-ax_padding,padding])

    var x_axis=d3.svg.axis().scale(x_scale);
    var y_axis=d3.svg.axis().scale(y_scale).orient('left');

    var svg=d3.select("#workarea svg");
    svg.append('g').attr('class','axis').attr('id','x_axis')
    .attr('transform','translate(0,'+(h - ax_padding)+')')
    .call(x_axis);
    svg.append('g').attr('class','axis').attr('id','y_axis')
    .attr('transform','translate('+ay_padding+',0)')
    .call(y_axis);

    points=svg.selectAll('circle').data(dataset).enter().append('circle')
    .attr('r',2)
    .attr('cx',function(d){return x_scale(d[0])})
    .attr('cy',function(d){return y_scale(d[1])})

}


d3_circles.js
/* 
 * Data Driven D3 Circles Example
 * Copyright SJ Hindmarch 2013, 2014
 * Distributed under the GNU General Public License
 */

function create(dataPoints){
    svg=d3.select("#svg");

    var circle=svg.selectAll("circle");
    circle.data(dataPoints).enter().append("circle")
    .attr("cx",function(d){return (2*d)})
    .attr("cy",function(d,i){return (50+10*i)})
    .attr("r",20)
    .style("fill","none");

    var text=svg.selectAll("text");
    text.data(dataPoints).enter().append("text")
    .text(function(d,i){return "("+d+","+i+")"})
    .attr("text-anchor","middle").attr("dy",".30em").attr("font-size","75%")
    .attr("x",function(d){return (2*d)})
    .attr("y",function(d,i){return (50+10*i)})
    ;

}

function cfill(colour){
    svg=d3.select("#svg");
    var circle=svg.selectAll("circle");
    circle.style("fill",colour);
}
	
function cyalign(state){
    svg=d3.select("#svg");
    var circle=svg.selectAll("circle");
    var text=svg.selectAll("text");
    if(state){
	circle.attr("cy",90);
	text.attr("y",90);
    }else{
	var p=function(d,i){return (50+10*i)};
	circle.attr("cy",p);
	text.attr("y",p);
    }
}

function cxalign(state){
    svg=d3.select("#svg");
    var circle=svg.selectAll("circle");
    var text=svg.selectAll("text");
    if(state){
	circle.attr("cx",180);
	text.attr("x",180);
    }else{
	var p=function(d,i){return (2*d)};
	circle.attr("cx",p);
	text.attr("x",p);
    }
}

function csize(state){
    svg=d3.select("#svg");
    var circle=svg.selectAll("circle");
    if(state){
	var a=function(d,i){return 5*Math.sqrt(d)};
	circle.attr("r",a);
    }else{
	circle.attr("r",20);
    }
}

function init(){
    var data=[10,25,60,94];
    create(data);
}


d3_iwd_v3.js
/* 
 * iWD Priority Visualisation With D3
 * Copyright SJ Hindmarch 2013, 2014
 * Distributed under the GNU General Public License
 *
 * Create priority objects by reading from a table
 *
 */

/* Create a profile function */
function DefaultProfileFunction(basePriority,baseWeight){
    this.name="default";
    this.basePriority=basePriority;
    this.baseWeight=baseWeight;
}

DefaultProfileFunction.prototype.getInitialPriority=function(bv){
    return this.basePriority+(bv*this.baseWeight);
};

DefaultProfileFunction.prototype.getTransitions=function(bv){
    return [];
};

DefaultProfileFunction.prototype.createTransitionsFromShape=function(shape,bv){
    var transitions=[];
    var base=this.basePriority;
    var weight=this.baseWeight;
    shape.forEach(function(x){
	    var t=x[0];
	    var p=x[1]+base+(bv*weight);
	    transitions.push([t,p]);
	});
    return transitions;
};

/* Subclass the profile functions */
function FastPeakProfileFunction(basePriority,baseWeight){
    DefaultProfileFunction.call(this,basePriority,baseWeight);
}
FastPeakProfileFunction.prototype=new DefaultProfileFunction();
FastPeakProfileFunction.constructor=DefaultProfileFunction;

function SlowPeakProfileFunction(basePriority,baseWeight){
    DefaultProfileFunction.call(this,basePriority,baseWeight);
}
SlowPeakProfileFunction.prototype=new DefaultProfileFunction();
SlowPeakProfileFunction.constructor=DefaultProfileFunction;

function HighTailProfileFunction(basePriority,baseWeight){
    DefaultProfileFunction.call(this,basePriority,baseWeight);
}
HighTailProfileFunction.prototype=new DefaultProfileFunction();
HighTailProfileFunction.constructor=DefaultProfileFunction;

function LowTailProfileFunction(basePriority,baseWeight){
    DefaultProfileFunction.call(this,basePriority,baseWeight);
}
LowTailProfileFunction.prototype=new DefaultProfileFunction();
LowTailProfileFunction.constructor=DefaultProfileFunction;

/* A fast peak reaches a peak quickly and then subsides slowly */
FastPeakProfileFunction.prototype.getTransitions=function(bv){
    this.shape=[[-30,1000],[5,800],[30,600],[60,400],[120,200]];
    return this.createTransitionsFromShape(this.shape,bv)
};
/* A slow peak reaches a peak slowly and then subsides quickly */
SlowPeakProfileFunction.prototype.getTransitions=function(bv){
    this.shape=[[-120,400],[-90,600],[-60,800],[-30,1000],[5,200]];
    return this.createTransitionsFromShape(this.shape,bv)
};
/* A high tail has long past past due item higher than due now */
HighTailProfileFunction.prototype.getTransitions=function(bv){
    this.shape=[[-30,800],[5,600],[60,1000]];
    return this.createTransitionsFromShape(this.shape,bv)
};
/* A low tail has long past past due item lower than due now */
LowTailProfileFunction.prototype.getTransitions=function(bv){
    this.shape=[[-30,1000],[5,600],[60,800]];
    return this.createTransitionsFromShape(this.shape,bv)
};

/* Create a child priority that uses a profile function to determine
 * transitions */
function ProfilePriority(name,color,profile,bv){
    this.basePriority=1000;
    this.baseWeight=1;

    var name=name;
    //console.log('ProfilePriority::constructor: '+name);
    var color=color;
    this.profile=profile;
    this.bv=+bv;
    this.profileFunction=this.createProfileFunction();
    var initP=this.profileFunction.getInitialPriority(this.bv);
    var transitions=this.profileFunction.getTransitions(this.bv);
    Priority.call(this,name,color,initP,transitions);
}

/* Set up inheritance */
ProfilePriority.prototype=new Priority();
ProfilePriority.constructor=ProfilePriority;

ProfilePriority.prototype.createProfileFunction=function(){
    //console.log('ProfilePriority::createProfileFunction: '+this.profile);
    switch(this.profile){
      case "fast peak":
	return new FastPeakProfileFunction(this.basePriority,this.baseWeight);
      case "slow peak":
	return new SlowPeakProfileFunction(this.basePriority,this.baseWeight);
      case "high tail":
	return new HighTailProfileFunction(this.basePriority,this.baseWeight);
      case "low tail":
	return new LowTailProfileFunction(this.basePriority,this.baseWeight);
      default:
	return new DefaultProfileFunction(this.basePriority,this.baseWeight);
    }
}

/* A Child of ProfilePriority that can be created from a row selector */
function ProfileRowPriority(row){
    this.row=row;
    var name=row.select('.name').text();
    var color=row.select('.color').text();
    var profile=row.select('.profile').node().value;
    var bv=row.select('.bv').node().value;
    ProfilePriority.call(this,name,color,profile,bv);
}

/* Set up inheritance */
ProfileRowPriority.prototype=new ProfilePriority();
ProfileRowPriority.constructor=ProfileRowPriority;

/* Create the selector for the profile */
function appendProfileSelector(row,profile){
    var profileNames=["default","fast peak","slow peak","high tail","low tail"];

    if(!profile){
	profile="default";
    }
    var select=row.append('select').attr('class','profile');
    profileNames.forEach(function(profileName){
	    var option=select.append('option').text(profileName);
	    if(profileName==profile){
		option.attr('selected','true');
	    }
	});
}  

/* Create the table from initial data */
function createProfilePrioritiesTable(tableId,priorities){
    var table=d3.select(tableId);
    priorities.forEach(function(priority){
	    var row=table.append('tr').attr('class','priority');
	    row.append('td').attr('class','name').text(priority.name);
	    row.append('td').attr('class','color').text(priority.color);
	    appendProfileSelector(row,priority.profile);
	    var bv=priority.bv;
	    if(!bv){
		bv=1000;
	    }
	    row.append('td').append('input').attr('class','bv')
	    .attr('value',bv);
	    row.append('td').text('\xA0'); // non-breaking space
	});

}

/* add a row to the table */
function addProfilePriorityRow(chart,tableId,formId){

    var form=d3.select(formId);
    var name=form.select('.name').node().value;
    var color=form.select('.color').node().value;
    var profile=form.select('.profile').node().value;
    var bv=form.select('.bv').node().value;
    
    var table=d3.select(tableId);
    var row=table.append('tr').attr('class','priority');
    row.append('td').attr('class','name').attr('name',name).text(name);
    row.append('td').attr('class','color').text(color);
    appendProfileSelector(row,profile);
    row.append('td').append('input').attr('class','bv').attr('value',bv);

    row.append('td').append('button').attr('type','button').text('Delete')
    .attr('onclick','deleteProfilePriorityRow(\''+chart+
	  '\',\''+tableId+'\',\''+name+'\')');
}

/* delete a row from the table */
function deleteProfilePriorityRow(chart,tableId,name){
    var cell=d3.select(tableId).selectAll('td[name='+name+']');
    cell[0].forEach(function(d,i){
	    d.parentElement.remove();
	});
}
